contracts/test/integration/IntegrationTestVanillaPSM.t.sol

```diff
@@ -0,0 +1,255 @@
+            endingPSMUnderlyingBalance
+        );
+    }
+
+    /// @notice redeem fails without approval
+    function testSwapVoltForUSDCFailsWithoutApproval() public {
+        vm.expectRevert(bytes("ERC20: transfer amount exceeds allowance"));
+
+        vanillaPsm.redeem(address(this), mintAmount, mintAmount / 1e12);
+    }
+
+    /// @notice redeem fails without approval
+    function testMintFailsWhenMintExceedsPSMBalance() public {
+        underlyingToken.approve(address(vanillaPsm), type(uint256).max);
+
+        uint256 currentPegPrice = oracle.getCurrentOraclePrice();
+        uint256 psmVoltBalance = volt.balanceOf(address(vanillaPsm));
+
+        // we get the amount we want to put in by getting the total PSM balance and dividing by the current peg price
+        // this lets us get the maximum amount we can deposit
+        uint256 amountIn = (psmVoltBalance * currentPegPrice) / 1e6;
+
+        // this will revert (correctly) as the math above is less precise than the PSMs, therefore our amountIn
+        // will slightly exceed the balance the PSM can give to us.
+        vm.expectRevert(
+            bytes("PegStabilityModule: Mint amount exceeds balance")
+        );
+
+        vanillaPsm.mint(address(this), amountIn, psmVoltBalance);
+    }
+
+    /// @notice mint fails without approval
+    function testSwapUnderlyingForVoltFailsWithoutApproval() public {
+        vm.expectRevert(bytes("ERC20: transfer amount exceeds allowance"));
+
+        vanillaPsm.mint(address(this), mintAmount, 0);
+    }
+
+    /// @notice withdraw erc20 fails without correct permissions
+    function testERC20WithdrawFailure() public {
+        vm.expectRevert(bytes("CoreRef: Caller is not a PCV controller"));
+
+        vanillaPsm.withdrawERC20(address(underlyingToken), address(this), 100);
+    }
+
+    /// @notice withdraw erc20 succeeds with correct permissions
+    function testERC20WithdrawSuccess() public {
+        vm.prank(MainnetAddresses.GOVERNOR);
+        core.grantPCVController(address(this));
+
+        uint256 startingBalance = underlyingToken.balanceOf(address(this));
+        vanillaPsm.withdrawERC20(
+            address(underlyingToken),
+            address(this),
+            mintAmount
+        );
+        uint256 endingBalance = underlyingToken.balanceOf(address(this));
+
+        assertEq(endingBalance - startingBalance, mintAmount);
+    }
+
+    /// @notice redeem fails when paused
+    function testRedeemFailsWhenPaused() public {
+        vm.prank(MainnetAddresses.GOVERNOR);
+        vanillaPsm.pause();
+
+        vm.expectRevert(bytes("Pausable: paused"));
+        vanillaPsm.redeem(address(this), 100, 100);
+    }
+
+    /// @notice mint fails when paused
+    function testMintFailsWhenPaused() public {
+        vm.prank(MainnetAddresses.GOVERNOR);
+        vanillaPsm.pause();
+
+        vm.expectRevert(bytes("Pausable: paused"));
+        vanillaPsm.mint(address(this), 100, 100);
+    }
+}
```

Comment:
would be good to add some fuzz tests here that prove equivalent behavior with a bunch of different inputs